README.md
sixad
=====

sixad can connect PS3 hardware (Sixaxis/DualShock3, Keypads and Remotes)
to a Linux-compatible machine.

Currently sixad supports:
 - Sixaxis/DualShock3 (buttons, axis, accelerometers, leds and rumble)
 - PS3 Navigation controller
 - PS3 Keypads
 - PS3 BD Remotes

sixad also registers the hardware's MAC/ID in the device name
(ex: "PLAYSTATION(R)3 Controller (00:XX:X0:0X:XX)".

-------------------------------------

sixad is copyright Filipe Coelho, licensed under the GNU GPL v2 license.


debian/changelog
sixad (1.6.0~vivid1) vivid; urgency=medium

  * Release to Vivid.

 -- Mike Purvis <mike@uwmike.com>  Fri, 21 Aug 2015 11:39:49 -0400

sixad (1.6.0~trusty2) trusty; urgency=medium

  * Add pkg-config to build depends.

 -- Mike Purvis <mike@uwmike.com>  Fri, 21 Aug 2015 11:23:48 -0400

sixad (1.6.0~trusty1) trusty; urgency=medium

  * Added short timeout for bluetooth transmissions.
  * Remove former 30 min activity timeout.

 -- Mike Purvis <mike@uwmike.com>  Fri, 21 Aug 2015 10:26:25 -0400

qtsixa (1.5.1+git20130130-0~trusty1) trusty; urgency=medium

  * Trusty build

 -- falkTX <falktx@gmail.com>  Sun, 20 Apr 2014 20:02:25 +0100

qtsixa (1.5.1+git20130130-0~saucy2) saucy; urgency=low

  * Saucy build

 -- falkTX <falktx@gmail.com>  Mon, 20 May 2013 11:26:46 +0100

qtsixa (1.5.1+git20130130-0~precise2) precise; urgency=low

  * Update to latest git

 -- falkTX <falktx@gmail.com>  Wed, 30 Jan 2013 20:09:31 +0000

qtsixa (1.5.1-0ubuntu1~lucid1) lucid; urgency=low

  * New upstream stable release

 -- falkTX <falktx@gmail.com>  Wed, 12 Oct 2011 17:50:42 +0100

qtsixa (1.4.97+git20110701-0~lucid1) lucid; urgency=low

  * New Git, remove hcid from install

 -- falkTX <falktx@gmail.com>  Fri, 01 Jul 2011 17:20:44 +0100

qtsixa (1.4.97+git20110520-0~lucid1) lucid; urgency=low

  * Update to latest GIT code

 -- falkTX <falktx@gmail.com>  Fri, 20 May 2011 22:11:48 +0100

qtsixa (1.4.97-1~natty3) natty; urgency=low

  * Natty build

 -- falkTX <falktx@gmail.com>  Fri, 21 Jan 2011 15:09:49 +0000

qtsixa (1.4.97-1~maverick3) maverick; urgency=low

  * Maverick build

 -- falkTX <falktx@gmail.com>  Fri, 21 Jan 2011 15:09:44 +0000

qtsixa (1.4.97-1~lucid3) lucid; urgency=low

  * Fix Natty build

 -- falkTX <falktx@gmail.com>  Fri, 21 Jan 2011 15:05:15 +0000

qtsixa (1.4.97-1~lucid2) lucid; urgency=low

  * Fixed neverball profile
  * sixad-jack improvements:
    - don't print messages (for performance)
    - better keys sensitivity

 -- falkTX <falktx@gmail.com>  Tue, 18 Jan 2011 23:55:46 +0000

qtsixa (1.4.97-1~lucid1) lucid; urgency=low

  * Implemented legacy mode
  * Fix segfault when config files have missing information

 -- falkTX <falktx@gmail.com>  Sun, 16 Jan 2011 22:55:54 +0000

qtsixa (1.4.96-1~lucid4) lucid; urgency=low

  * Disable sixad-raw on GUI

 -- falkTX <falktx@gmail.com>  Thu, 30 Dec 2010 16:50:02 +0000

qtsixa (1.4.96-1~lucid3) lucid; urgency=low

  * Several fixes for sixad-raw

 -- falkTX <falktx@gmail.com>  Thu, 30 Dec 2010 01:09:34 +0000

qtsixa (1.4.96-0~lucid2) lucid; urgency=low

  * Fix Scroll usage
  * Fix using keys in the analog sticks
  * Prepare package for both Lucid, Maverick and Natty builds

 -- falkTX <falktx@gmail.com>  Thu, 23 Dec 2010 00:27:58 +0000

qtsixa (1.4.95-0~ppa6) lucid; urgency=low

  * Implemented log-rotation for sixad logs

 -- falkTX <falktx@gmail.com>  Fri, 03 Dec 2010 17:32:12 +0000

qtsixa (1.4.95-0~ppa2) lucid; urgency=low

  * Implemented lr3 axis/buttons thing
  * Fixed manage devices list sometimes empty
  * Fix Scroll speed (was too fast)

 -- falkTX <falktx@gmail.com>  Wed, 01 Dec 2010 15:40:03 +0000

qtsixa (1.4.95-0~ppa1) lucid; urgency=low

  * Re-write of backend code in C++ (many fixes, working rumble support)
  * Split sixad and utils
  * Update GUI and other stuff where needed

 -- falkTX <falktx@gmail.com>  Sat, 20 Nov 2010 20:58:28 +0000

qtsixa (1.4.90) lucid; urgency=low

  * Initial release. (Closes: #XXXXXX)

 -- falkTX <falktx@gmail.com>  Tue, 25 May 2010 18:08:08 +0100


shared.h
/*
 * shared.h
 *
 * This file is part of the QtSixA, the Sixaxis Joystick Manager
 * Copyright 2008-2011 Filipe Coelho <falktx@gmail.com>
 *
 * QtSixA can be redistributed and/or modified under the terms of the GNU General
 * Public License (Version 2), as published by the Free Software Foundation.
 * A copy of the license is included in the QtSixA source code, or can be found
 * online at www.gnu.org/licenses.
 *
 * QtSixA is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 */

#ifndef SHARED_H
#define SHARED_H

#include <unistd.h>

struct dev_led {
    bool enabled;
    bool anim;
    bool auto_sel;
    int number;
};

struct dev_joystick {
    bool enabled;
    bool buttons;
    bool axis;
    bool sbuttons;
    bool accel;
    bool accon;
    bool speed;
    bool pos;
};

struct dev_remote {
    bool enabled;
    bool numeric;
    bool dvd;
    bool directional;
    bool multimedia;
};

struct dev_input {
    bool enabled;
    int key_select, key_l3, key_r3, key_start, key_up, key_right, key_down, key_left;
    int key_l2, key_r2, key_l1, key_r1, key_tri, key_cir, key_squ, key_cro, key_ps;
    int axis_l_type, axis_r_type, axis_speed;
    int axis_l_up, axis_l_right, axis_l_down, axis_l_left;
    int axis_r_up, axis_r_right, axis_r_down, axis_r_left;
    bool use_lr3;
};

struct dev_rumble {
    bool enabled;
    bool old_mode;
};

struct dev_safety_timeout {
    bool enabled;
    int timeout_ms;
};

struct device_settings {
    struct dev_led led;
    struct dev_joystick joystick;
    struct dev_remote remote;
    struct dev_input input;
    struct dev_rumble rumble;
    struct dev_safety_timeout safety_timeout;
};

bool was_active();
void set_active(int active);

bool io_canceled();
void sig_term(int sig);
void open_log(const char *app_name);

struct device_settings init_values(const char *mac);

int get_joystick_number();
void enable_sixaxis(int csk);

#endif // SHARED_H


shared.cpp
/*
 * shared.cpp
 *
 * This file is part of the QtSixA, the Sixaxis Joystick Manager
 * Copyright 2008-2011 Filipe Coelho <falktx@gmail.com>
 *
 * QtSixA can be redistributed and/or modified under the terms of the GNU General
 * Public License (Version 2), as published by the Free Software Foundation.
 * A copy of the license is included in the QtSixA source code, or can be found
 * online at www.gnu.org/licenses.
 *
 * QtSixA is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 */

#include "shared.h"
#include "textfile.h"

#include <cstdio>
#include <cstring>
#include <fcntl.h>
#include <signal.h>
#include <syslog.h>
#include <unistd.h>
#include <sys/socket.h>

static volatile sig_atomic_t __active = false;
static volatile sig_atomic_t __io_canceled = false;

bool was_active()
{
  return __active;
}

void set_active(int active)
{
  __active = active;
}

bool io_canceled()
{
  return __io_canceled;
}

void sig_term(int /* sig */)
{
    __io_canceled = 1;
}

void open_log(const char *app_name)
{
    openlog(app_name, LOG_NDELAY|LOG_PID|LOG_PERROR, LOG_DAEMON);
}

struct device_settings init_values(const char *addr)
{
    struct device_settings settings;

    char pathname[64];
    strcpy(pathname, "/var/lib/sixad/profiles/");
    strcat(pathname, addr);

    if (open(pathname, O_RDONLY) > 0) {

        settings.led.enabled = textfile_get_int(pathname, "enable_leds", 1);
        settings.led.auto_sel = textfile_get_int(pathname, "led_n_auto", 1);
        settings.led.number = textfile_get_int(pathname, "led_n_number", 1);
        settings.led.anim = textfile_get_int(pathname, "led_anim", 1);

        settings.joystick.enabled = textfile_get_int(pathname, "enable_joystick", 1);
        settings.joystick.buttons = textfile_get_int(pathname, "enable_buttons", 1);
        settings.joystick.sbuttons = textfile_get_int(pathname, "enable_sbuttons", 1);
        settings.joystick.axis  = textfile_get_int(pathname, "enable_axis", 1);
        settings.joystick.accel = textfile_get_int(pathname, "enable_accel", 1);
        settings.joystick.accon = textfile_get_int(pathname, "enable_accon", 0);
        settings.joystick.speed = textfile_get_int(pathname, "enable_speed", 0);
        settings.joystick.pos = textfile_get_int(pathname, "enable_pos", 0);

        settings.remote.enabled = textfile_get_int(pathname, "enable_remote", 1);
        settings.remote.numeric = textfile_get_int(pathname, "remote_numberic", 1);
        settings.remote.dvd = textfile_get_int(pathname, "remote_dvd", 1);
        settings.remote.directional = textfile_get_int(pathname, "remote_directional", 1);
        settings.remote.multimedia = textfile_get_int(pathname, "remote_multimedia", 1);

        settings.input.enabled = textfile_get_int(pathname, "enable_input", 0);
        settings.input.key_select = textfile_get_int(pathname, "key_select", 0);
        settings.input.key_l3 = textfile_get_int(pathname, "key_l3", 0);
        settings.input.key_r3 = textfile_get_int(pathname, "key_r3", 0);
        settings.input.key_start = textfile_get_int(pathname, "key_start", 0);
        settings.input.key_up    = textfile_get_int(pathname, "key_up", 0);
        settings.input.key_right = textfile_get_int(pathname, "key_right", 0);
        settings.input.key_down  = textfile_get_int(pathname, "key_down", 0);
        settings.input.key_left  = textfile_get_int(pathname, "key_left", 0);
        settings.input.key_l2  = textfile_get_int(pathname, "key_l2", 0);
        settings.input.key_r2  = textfile_get_int(pathname, "key_r2", 0);
        settings.input.key_l1  = textfile_get_int(pathname, "key_l1", 0);
        settings.input.key_r1  = textfile_get_int(pathname, "key_r1", 0);
        settings.input.key_tri = textfile_get_int(pathname, "key_tri", 0);
        settings.input.key_cir = textfile_get_int(pathname, "key_cir", 0);
        settings.input.key_squ = textfile_get_int(pathname, "key_squ", 0);
        settings.input.key_cro = textfile_get_int(pathname, "key_cro", 0);
        settings.input.key_ps  = textfile_get_int(pathname, "key_ps", 0);
        settings.input.axis_l_type  = textfile_get_int(pathname, "axis_left_type", 0);
        settings.input.axis_l_up    = textfile_get_int(pathname, "axis_left_up", 0);
        settings.input.axis_l_right = textfile_get_int(pathname, "axis_left_right", 0);
        settings.input.axis_l_down  = textfile_get_int(pathname, "axis_left_down", 0);
        settings.input.axis_l_left  = textfile_get_int(pathname, "axis_left_left", 0);
        settings.input.axis_r_type  = textfile_get_int(pathname, "axis_right_type", 0);
        settings.input.axis_r_up    = textfile_get_int(pathname, "axis_right_up", 0);
        settings.input.axis_r_right = textfile_get_int(pathname, "axis_right_right", 0);
        settings.input.axis_r_down  = textfile_get_int(pathname, "axis_right_down", 0);
        settings.input.axis_r_left  = textfile_get_int(pathname, "axis_right_left", 0);
        settings.input.axis_speed = (10 - textfile_get_int(pathname, "axis_speed", 6));
        settings.input.use_lr3 = textfile_get_int(pathname, "use_lr3", 0);

        settings.rumble.enabled = textfile_get_int(pathname, "enable_rumble", 1);
        settings.rumble.old_mode = textfile_get_int(pathname, "old_rumble_mode", 0);

        settings.safety_timeout.enabled = textfile_get_int(pathname, "enable_safety_timeout", 1);
        settings.safety_timeout.timeout_ms = textfile_get_int(pathname, "safety_timeout_ms", 250);

    } else if (open("/var/lib/sixad/profiles/default", O_RDONLY) > 0) { //default config
        strcpy(pathname, "/var/lib/sixad/profiles/default");

        settings.led.enabled = textfile_get_int(pathname, "enable_leds", 1);
        settings.led.auto_sel = textfile_get_int(pathname, "led_n_auto", 1);
        settings.led.number = textfile_get_int(pathname, "led_n_number", 1);
        settings.led.anim = textfile_get_int(pathname, "led_anim", 1);

        settings.joystick.enabled = textfile_get_int(pathname, "enable_joystick", 1);
        settings.joystick.buttons = textfile_get_int(pathname, "enable_buttons", 1);
        settings.joystick.sbuttons = textfile_get_int(pathname, "enable_sbuttons", 1);
        settings.joystick.axis  = textfile_get_int(pathname, "enable_axis", 1);
        settings.joystick.accel = textfile_get_int(pathname, "enable_accel", 1);
        settings.joystick.accon = textfile_get_int(pathname, "enable_accon", 0);
        settings.joystick.speed = textfile_get_int(pathname, "enable_speed", 0);
        settings.joystick.pos = textfile_get_int(pathname, "enable_pos", 0);

        settings.remote.enabled = textfile_get_int(pathname, "enable_remote", 1);
        settings.remote.numeric = textfile_get_int(pathname, "remote_numberic", 1);
        settings.remote.dvd = textfile_get_int(pathname, "remote_dvd", 1);
        settings.remote.directional = textfile_get_int(pathname, "remote_directional", 1);
        settings.remote.multimedia = textfile_get_int(pathname, "remote_multimedia", 1);

        settings.input.enabled = textfile_get_int(pathname, "enable_input", 0);
        settings.input.key_select = textfile_get_int(pathname, "key_select", 0);
        settings.input.key_l3 = textfile_get_int(pathname, "key_l3", 0);
        settings.input.key_r3 = textfile_get_int(pathname, "key_r3", 0);
        settings.input.key_start = textfile_get_int(pathname, "key_start", 0);
        settings.input.key_up    = textfile_get_int(pathname, "key_up", 0);
        settings.input.key_right = textfile_get_int(pathname, "key_right", 0);
        settings.input.key_down  = textfile_get_int(pathname, "key_down", 0);
        settings.input.key_left  = textfile_get_int(pathname, "key_left", 0);
        settings.input.key_l2  = textfile_get_int(pathname, "key_l2", 0);
        settings.input.key_r2  = textfile_get_int(pathname, "key_r2", 0);
        settings.input.key_l1  = textfile_get_int(pathname, "key_l1", 0);
        settings.input.key_r1  = textfile_get_int(pathname, "key_r1", 0);
        settings.input.key_tri = textfile_get_int(pathname, "key_tri", 0);
        settings.input.key_cir = textfile_get_int(pathname, "key_cir", 0);
        settings.input.key_squ = textfile_get_int(pathname, "key_squ", 0);
        settings.input.key_cro = textfile_get_int(pathname, "key_cro", 0);
        settings.input.key_ps  = textfile_get_int(pathname, "key_ps", 0);
        settings.input.axis_l_type  = textfile_get_int(pathname, "axis_left_type", 0);
        settings.input.axis_l_up    = textfile_get_int(pathname, "axis_left_up", 0);
        settings.input.axis_l_right = textfile_get_int(pathname, "axis_left_right", 0);
        settings.input.axis_l_down  = textfile_get_int(pathname, "axis_left_down", 0);
        settings.input.axis_l_left  = textfile_get_int(pathname, "axis_left_left", 0);
        settings.input.axis_r_type  = textfile_get_int(pathname, "axis_right_type", 0);
        settings.input.axis_r_up    = textfile_get_int(pathname, "axis_right_up", 0);
        settings.input.axis_r_right = textfile_get_int(pathname, "axis_right_right", 0);
        settings.input.axis_r_down  = textfile_get_int(pathname, "axis_right_down", 0);
        settings.input.axis_r_left  = textfile_get_int(pathname, "axis_right_left", 0);
        settings.input.axis_speed = (10 - textfile_get_int(pathname, "axis_speed", 6));
        settings.input.use_lr3 = textfile_get_int(pathname, "use_lr3", 0);

        settings.rumble.enabled = textfile_get_int(pathname, "enable_rumble", 1);
        settings.rumble.old_mode = textfile_get_int(pathname, "old_rumble_mode", 0);

        settings.safety_timeout.enabled = textfile_get_int(pathname, "enable_safety_timeout", 1);
        settings.safety_timeout.timeout_ms = textfile_get_int(pathname, "safety_timeout_ms", 250);
    } else { // no config

        settings.led.enabled = 1;
        settings.led.auto_sel = 1;
        settings.led.number = 1;
        settings.led.anim = 1;

        settings.joystick.enabled = 1;
        settings.joystick.buttons = 1;
        settings.joystick.sbuttons = 1;
        settings.joystick.axis  = 1;
        settings.joystick.accel = 1;
        settings.joystick.accon = 0;
        settings.joystick.speed = 0;
        settings.joystick.pos = 0;

        settings.remote.enabled = 1;
        settings.remote.numeric = 1;
        settings.remote.dvd = 1;
        settings.remote.directional = 1;
        settings.remote.multimedia = 1;

        settings.input.enabled = 0;
        settings.input.key_select = 0;
        settings.input.key_l3 = 0;
        settings.input.key_r3 = 0;
        settings.input.key_start = 0;
        settings.input.key_up    = 0;
        settings.input.key_right = 0;
        settings.input.key_down  = 0;
        settings.input.key_left  = 0;
        settings.input.key_l2  = 0;
        settings.input.key_r2  = 0;
        settings.input.key_l1  = 0;
        settings.input.key_r1  = 0;
        settings.input.key_tri = 0;
        settings.input.key_cir = 0;
        settings.input.key_squ = 0;
        settings.input.key_cro = 0;
        settings.input.key_ps  = 0;
        settings.input.axis_l_type = 0;
        settings.input.axis_l_up = 0;
        settings.input.axis_l_right = 0;
        settings.input.axis_l_down = 0;
        settings.input.axis_l_left = 0;
        settings.input.axis_r_type = 0;
        settings.input.axis_r_up = 0;
        settings.input.axis_r_right = 0;
        settings.input.axis_r_down = 0;
        settings.input.axis_r_left = 0;
        settings.input.axis_speed = 6;
        settings.input.use_lr3 = 0;

        settings.rumble.enabled = 1;
        settings.rumble.old_mode = 0;

        settings.safety_timeout.enabled = 1;
        settings.safety_timeout.timeout_ms = 250;
    }

    return settings;
}

int get_joystick_number()
{
    int i, js;
    char jspath[16];

    js = 1;
    for (i=0; i < 10; i++) {
        snprintf(jspath, sizeof(jspath), "/dev/input/js%i", i);
        if (open(jspath, O_RDONLY) > 0) {
            js = i+1;
        }
    }

    return js;
}

void enable_sixaxis(int csk)
{
#ifdef GASIA_GAMEPAD_HACKS
    unsigned char enable[] = {
        0xA2,
        0x01,
        0x00, 0x00, 0x00, 0x00, 0x00,   // rumble values [0x00, right-timeout, right-force, left-timeout, left-force]
        0x00, 0x00, 0x00, 0x00, 0x02,   // 0x02=LED1 .. 0x10=LED4
        0xff, 0x27, 0x10, 0x00, 0x32,   // LED 4
        0xff, 0x27, 0x10, 0x00, 0x32,   // LED 3
        0xff, 0x27, 0x10, 0x00, 0x32,   // LED 2
        0xff, 0x27, 0x10, 0x00, 0x32,   // LED 1
        0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00, 0x00, 0x00,
        0x00, 0x00, 0x00
    };
#else
    char buf[128];
    unsigned char enable[] = {
        0x53, /* HIDP_TRANS_SET_REPORT | HIDP_DATA_RTYPE_FEATURE */
        0xf4, 0x42, 0x03, 0x00, 0x00
    };
#endif

    /* enable reporting */
    send(csk, enable, sizeof(enable), 0);
#ifndef GASIA_GAMEPAD_HACKS
    recv(csk, buf, sizeof(buf), 0);
#endif
}


sixad-sixaxis.cpp
/*
 * sixad-sixaxis.cpp
 *
 * This file is part of the QtSixA, the Sixaxis Joystick Manager
 * Copyright 2008-10 Filipe Coelho <falktx@gmail.com>
 *
 * QtSixA can be redistributed and/or modified under the terms of the GNU General
 * Public License (Version 2), as published by the Free Software Foundation.
 * A copy of the license is included in the QtSixA source code, or can be found
 * online at www.gnu.org/licenses.
 *
 * QtSixA is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 */

#include "shared.h"
#include "sixaxis.h"
#include "uinput.h"

#include <cstdlib>
#include <iostream>
#include <poll.h>
#include <signal.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <sys/socket.h>

int csk = 0;
int isk = 1;
int debug;

int led_n;
bool old_rumble_mode;

volatile bool active = false;
volatile int weak = 0;
volatile int strong = 0;
volatile int timeout = 0;

struct uinput_fd *ufd;

static void uinput_listen()
{
        struct input_event event;
        struct uinput_ff_upload upload;
        struct uinput_ff_erase erase;
        struct rumble_effect effects[MAX_RUMBLE_EFFECTS];
        int i;

        //reset effects
        for (i=0; i<MAX_RUMBLE_EFFECTS; i++) {
            effects[i].id = -2;
        }

        while (!io_canceled()) {
                if (read(ufd->js, &event, sizeof event) != sizeof event) {
                        syslog(LOG_INFO, "uinput_listen::Error on uinput read");
                        continue;
                }

                if (debug) syslog(LOG_INFO, "GOT event :: type is %i; code is %i; value is %i\n", event.type, event.code, event.value);

                switch (event.type) {
                case EV_UINPUT:
                        switch (event.code) {
                        case UI_FF_UPLOAD:
                                memset(&upload, 0, sizeof(upload));
                                upload.request_id = event.value;
                                if (ioctl(ufd->js, UI_BEGIN_FF_UPLOAD, &upload) < 0) {
                                        syslog(LOG_ERR, "uinput_listen::Error on ff upload begin");
                                }
                                weak = upload.effect.u.rumble.weak_magnitude/0x101;
                                strong = upload.effect.u.rumble.strong_magnitude/0x101;
                                timeout = upload.effect.replay.length/0x101;
                                if (!old_rumble_mode) {
                                    for (i=0; i<MAX_RUMBLE_EFFECTS; i++) {
                                        if (effects[i].id == -2) {
                                            effects[i].id = upload.effect.id;
                                            effects[i].weak = weak;
                                            effects[i].strong = strong;
                                            effects[i].timeout = timeout;
                                            if (debug) syslog(LOG_INFO, "going for effect #%i", i);
                                            break;
                                        }
                                    }
                                } else {
                                    active = true;
                                }
                                upload.retval = 0;
                                if (ioctl(ufd->js, UI_END_FF_UPLOAD, &upload) < 0) {
                                        syslog(LOG_ERR, "uinput_listen::Error on ff upload end");
                                }
                                break;
                        case UI_FF_ERASE:
                                memset(&erase, 0, sizeof(erase));
                                erase.request_id = event.value;
                                if (ioctl(ufd->js, UI_BEGIN_FF_ERASE, &erase) < 0) {
                                        syslog(LOG_ERR, "uinput_listen::Error on ff erase begin");
                                }
                                if (!old_rumble_mode) {
                                    for (i=0; i<MAX_RUMBLE_EFFECTS; i++) {
                                        if (effects[i].id == event.value) {
                                            effects[i].id = -2;
                                            break;
                                        }
                                    }
                                } else {
                                    weak = 0;
                                    strong = 0;
                                    timeout = 0;
                                    active = true;
                                }
                                erase.retval = 0;
                                if (ioctl(ufd->js, UI_END_FF_ERASE, &erase) < 0) {
                                        syslog(LOG_ERR, "uinput_listen::Error on ff erase end");
                                }
                                break;
                        default:
                                break;
                        }
                        break;
                case EV_FF:
                        if (event.value) {
                            for (i=0; i<MAX_RUMBLE_EFFECTS; i++) {
                                if (effects[i].id == event.code) {
                                    do_rumble(csk, led_n, effects[i].weak, effects[i].strong, effects[i].timeout);
                                    if (debug) syslog(LOG_INFO, "RUMBLE now :: %i|%i|%i", effects[i].weak, effects[i].strong, effects[i].timeout);
                                    break;
                                }
                            }
                        } else {
                           if (!old_rumble_mode) {
                                do_rumble(csk, led_n, 0, 0, 0);
                                if (debug) syslog(LOG_INFO, "RUMBLE clean");
                           }
                        }
                        break;
                default:
                        break;
                }
        }

        pthread_exit((void*)1);
}

static void rumble_listen()
{
    while (!io_canceled()) {
        if (active) {
            do_rumble(csk, led_n, weak, strong, timeout);
            active = false;
        } else
            usleep(5000);
    }

    pthread_exit((void*)1);
}

static void process_sixaxis(struct device_settings settings, const char *mac)
{
    int br;
    bool msg = true;
    unsigned char buf[128];

    sigset_t sigs;
    sigfillset(&sigs);
    sigdelset(&sigs, SIGCHLD);
    sigdelset(&sigs, SIGPIPE);
    sigdelset(&sigs, SIGTERM);
    sigdelset(&sigs, SIGINT);
    sigdelset(&sigs, SIGHUP);

    while (!io_canceled()) {
        br = read(isk, buf, sizeof(buf));

        if (msg) {
            syslog(LOG_INFO, "Connected 'PLAYSTATION(R)3 Controller (%s)' [Battery %02X]", mac, buf[31]);
            msg = false;
        }

        if (br < 0) {
            break;
        } else if (br==50 && buf[0]==0xa1 && buf[1]==0x01 && buf[2]==0x00) { //only continue if we've got a Sixaxis
            if (settings.joystick.enabled) do_joystick(ufd->js, buf, settings.joystick);
            if (settings.input.enabled) do_input(ufd->mk, buf, settings.input);
        } else if (br==50 && buf[0]==0xa1 && buf[1]==0x01 && buf[2]==0xff) {
            if (debug) syslog(LOG_ERR, "Got 0xff Sixaxis buffer, ignored");
        } else if (buf[0]==0xa1 && buf[1]==0x01 && buf[2]==0x00) {
            syslog(LOG_ERR, "Bad Sixaxis buffer (out of battery?), disconnecting now...");
	          sig_term(0);
            break;
        } else {
            if (debug) syslog(LOG_ERR, "Non-Sixaxis packet received and ignored (0x%02x|0x%02x|0x%02x)", buf[0], buf[1], buf[2]);
        }

        // Use poll to wait on the bluetooth socket's next packet,
        // in order to enforce a strict timeout.
        if (settings.safety_timeout.enabled)
        {
            pollfd p;
            p.events = POLLIN;
            p.fd = isk;
            p.revents = 0;
            timespec timeout;
            timeout.tv_sec = 0;
            timeout.tv_nsec = 1000000 * settings.safety_timeout.timeout_ms;

            if (ppoll(&p, 1, &timeout, &sigs) <= 0)
            {
                syslog(LOG_ERR, "Controller connection timed out. Please re-pair.");

                // Clear all buttons to zero state and send a final update.
                buf[3] = buf[4] = buf[5] = 0;
                if (settings.joystick.enabled) do_joystick(ufd->js, buf, settings.joystick);
                if (settings.input.enabled) do_input(ufd->mk, buf, settings.input);

                // Disconnect.
                sig_term(0);
                break;
            }
        }
    }

    if (debug) syslog(LOG_ERR, "Read loop was broken on the Sixaxis process");
}

int main(int argc, char *argv[])
{
    struct pollfd p[4];
    struct timespec timeout;
    struct device_settings settings;
    struct sigaction sa;
    pthread_t uinput_listen_thread, rumble_listen_thread;
    sigset_t sigs;
    short events;

    if (argc < 3) {
        std::cout << "Running " << argv[0] << " requires 'sixad'. Please run sixad instead" << std::endl;
        return 1;
    }

    const char *mac = argv[1];
    debug = atoi(argv[2]);

    open_log("sixad-sixaxis");
    syslog(LOG_INFO, "started");
    settings = init_values(mac);
    settings.remote.enabled = false;

    ufd = uinput_open(DEV_TYPE_SIXAXIS, mac, settings);

    if (ufd->js < 0 || ufd->mk < 0) {
        return 1;
    } else if (ufd->js == 0 && ufd->mk == 0) {
        syslog(LOG_ERR, "sixaxis config has no joystick or input mode selected - please choose one!");
        return 1;
    }

    enable_sixaxis(csk);
    led_n = set_sixaxis_led(csk, settings.led, settings.rumble.enabled);

    if (settings.rumble.enabled) {
      old_rumble_mode = settings.rumble.old_mode;
      if (pthread_create(&uinput_listen_thread, NULL, (void *(*)(void *))uinput_listen, NULL)) {
              syslog(LOG_ERR, "error starting uinput listen thread");
              return 1;
      }
      if (settings.rumble.old_mode) {
          if (pthread_create(&rumble_listen_thread, NULL, (void *(*)(void *))rumble_listen, NULL)) {
              syslog(LOG_ERR, "error starting rumble listen thread");
              return 1;
          }
      }
    }

    sigfillset(&sigs);
    sigdelset(&sigs, SIGCHLD);
    sigdelset(&sigs, SIGPIPE);
    sigdelset(&sigs, SIGTERM);
    sigdelset(&sigs, SIGINT);
    sigdelset(&sigs, SIGHUP);

    memset(&sa, 0, sizeof(sa));
    sa.sa_flags = SA_NOCLDSTOP;

    sa.sa_handler = sig_term;
    sigaction(SIGTERM, &sa, NULL);
    sigaction(SIGINT, &sa, NULL);

    sa.sa_handler = SIG_IGN;
    sigaction(SIGCHLD, &sa, NULL);
    sigaction(SIGPIPE, &sa, NULL);

    p[0].events = POLLIN | POLLERR | POLLHUP;
    p[1].events = POLLIN | POLLERR | POLLHUP;
    p[2].events = POLLIN | POLLERR | POLLHUP;
    p[3].events = POLLIN | POLLERR | POLLHUP;

    p[0].fd = 0;
    p[1].fd = 1;
    p[2].fd = ufd->js;
    p[3].fd = ufd->mk;

    if (!ufd->js)
      p[2].fd = ufd->mk;

    int idx = (ufd->js && ufd->mk) ? 4 : 3;

    while (!io_canceled()) {
        int i;

        for (i = 0; i < 4; i++)
            p[i].revents = 0;

        timeout.tv_sec = 1;
        timeout.tv_nsec = 0;

        if (ppoll(p, idx, &timeout, &sigs) < 1)
            continue;

        if (p[1].revents & POLLIN) {
            process_sixaxis(settings, mac);
        }

        events = p[0].revents | p[1].revents | p[2].revents | p[3].revents;

        if (events & (POLLERR | POLLHUP)) {
            sig_term(0);
            break;
        }
    }

    if (settings.rumble.enabled) {
      if (pthread_cancel(uinput_listen_thread)) {
        if (pthread_join(uinput_listen_thread, NULL)) {
          syslog(LOG_ERR, "Error canceling uinput listen thread");
        }
      }
      if (settings.rumble.old_mode) {
          if (pthread_cancel(rumble_listen_thread)) {
            if (pthread_join(rumble_listen_thread, NULL)) {
              syslog(LOG_ERR, "Error canceling rumble listen thread");
            }
          }
      }
    }

    if (debug) syslog(LOG_INFO, "Closing uinput...");

    if (settings.joystick.enabled) {
        uinput_close(ufd->js, debug);
    }
    if (settings.input.enabled) {
        uinput_close(ufd->mk, debug);
    }

    do_rumble(csk, 10, 0xff, 0xff, 0x01);
    usleep(10*1000);

    shutdown(isk, SHUT_RDWR);
    shutdown(csk, SHUT_RDWR);

    delete ufd;

    // hack for force disconnect
    char cmd[32] = { 0 };
    strcpy(cmd, "hcitool dc ");
    strcat(cmd, mac);

    usleep(10*1000);
    syslog(LOG_INFO, "Force disconnect of \"%s\"", mac);
    system(cmd);

    if (debug) syslog(LOG_INFO, "Done");

    return 0;
}
